### PR TITLE
Fix SDPA rewriter 12 test pattern

### LINKED ISSUE #135407: [Inductor][SDPA] `test_sdpa_rewriter_12` broken on A2/A16 GPU
### 🐛 Describe the bug

One of the pattern-matching tests fails on A2/A16 and likely A10 (untested). Not particularly urgent but I would like to learn more about how to expose which part of Inductor is failing to pattern match on certain cases. Since I suspect this is using pre-compiled FA kernels, my guess is there is a hidden constraint that isn't met somewhere (not in the pattern definition itself) or there is a runtime failure that is caught and suppressed.

For working configs it seems to match to pattern 11 (not 12):
https://github.com/pytorch/pytorch/blob/a6b9d444fbd2ddcf8481ea6adbac08de5443f1fe/torch/_inductor/fx_passes/fuse_attention.py#L277

Is there a way to get the Inductor logs that would be relevant here?

CC @drisspg

### Versions

~week old main branch

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184295

Update the SDPA rewriter 12 test so the dropout is applied to attention weights before the value matmul, matching the actual pattern being exercised. Add an exact per-pattern assertion for the training case to catch regressions.

Fixes #135407

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — APPROVED** [review]

**Review by eellison — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]
